### PR TITLE
Feature/root history

### PR DESCRIPTION
**Pull request by imeckler**
This implements the root history data structured as described in the RFC

### CODE REVIEW

File: src/lib/root_history/root_history.mli [hunk 1]
```diff
@@ -0,0 +1,9 @@
+type 'a t
+
+val create : max_size:int -> 'a t
+
+val push : 'a t -> length:Coda_numbers.Length.t -> data:'a -> unit
+
+val find : 'a t -> Coda_numbers.Length.t -> 'a option
+
+val find_exn : 'a t -> Coda_numbers.Length.t -> 'a
```
Comment (bkase) [inline src/lib/root_history/root_history.mli]: can you doccomment the mli?

File: src/lib/root_history/root_history.ml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  remove_old_data length t
+
+let find t length =
+  Option.map (Hashtbl.find t.by_length length) ~f:(fun e ->
+      snd (Doubly_linked.Elt.value e) )
+
+let find_exn t length = Option.value_exn (find t length)
+
+let create ~max_size =
+  {max_size; queue= Doubly_linked.create (); by_length= Length.Table.create ()}
```
Comment (bkase) [inline src/lib/root_history/root_history.ml]: Consider writing a unit test -- maybe a quickcheck that pushes random lengths and assert that size is bounded by max-size

File: src/lib/root_history/root_history.mli [hunk 3]
```diff
@@ -1,9 +1,27 @@
 type 'a t
 
+(* This is a FIFO queue of values of type ['a], keyed by their length.
```
Comment (bkase) [inline src/lib/root_history/root_history.mli]: use `(**` for all of these and they'll render better

File: src/lib/root_history/root_history.mli [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+val push :
+     'a t
+  -> length:Coda_numbers.Length.t
+  -> data:'a
+  -> [`Ok | `Length_did_not_increase]
+
+val push_exn : 'a t -> length:Coda_numbers.Length.t -> data:'a -> unit
+
+(* Find an element by its length *)
+val find : 'a t -> Coda_numbers.Length.t -> 'a option
```
[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/root_history/root_history.mli]: oh right -- sorry about that

**Pull request merged**